public/app/features/dashboard/state/DashboardModel.ts

```diff
@@ -436,9 +437,17 @@ export class DashboardModel {
       return this.panelInEdit;
     }
 
-    for (const panel of this.panels) {
-      if (panel.id === id) {
-        return panel;
+    const panel = this.panels.find((p) => p.id === id);
+
+    if (panel) {
+      return panel;
+    }
+
+    const rows = this.panels.filter((p) => p.type === 'row');
+    for (const row of rows) {
+      const rowPanel: any = row.panels?.find((p: any) => p.id === id);
+      if (rowPanel) {
+        return new PanelModel(rowPanel);
```

Comment:
feels like these 3 loops can be combined (into a nested loolp)? instead of doing 1 loop overall panels, then another to find rows, and then a 3rd to loop over rows + nested loop over row panels.

Reply:
The number of loops could be reduced, but I thought having it this way was an okay compromise for the clarity it provides. I deliberately want rows to be dealt with after regular panels as otherwise new panel models might be created when it's unnecessary.

Reply:
>  I deliberately want rows to be dealt with after regular panels as otherwise new panel models might be created when it's unnecessary. 

When would this happen?

Reply:
Part of this PR is the idea of returning a new PanelModel from `getPanelById` if the panel with the provided ID exists, but is not otherwise a top-level panel (i.e. It is contained in a collapsed row). If I just iterated over `this.panels` and returned a new `PanelModel` if the panel with the provided ID was found in a row, if that row were expanded, that panel would also be contained later in `this.panels` which we could have returned, but instead we've made another PanelModel. Hopefully that explains it but if not maybe it would be easier to communicate by sharing my screen or something. 😅

Reply:
not that does not make much sense, a expanded row has an empty panels array (or null cannot remember), so if you loop through all panels and if it's a row panel loop through panel.panels then and find the panel there you would never have found it in the top level dashboard.panels array

Reply:
Sorry for rows, the way they are implemented does not make much sense :(  My fault, should have kept the nested structure, so that the nested structure is there always, no matter if the row is expanded or not

Reply:
You're right of course, just needed to make sure the panels were getting refreshed _after_ `panel.panels` gets set to `[]`